### PR TITLE
Add kube2iam to control AWS IAM policy access

### PR DESCRIPTION
**Pull request by icereval**
A [kube2iam](https://github.com/jtblin/kube2iam) chart.

Provides IAM credentials to pods running inside a AWS kubernetes cluster based on annotations.

Pods that accept annotations can run into the following issue:  https://github.com/kubernetes/helm/issues/1540

One workaround is to simply set the annotations directly in your charts `values.yaml` file.

@jmcarp

Based on https://github.com/18F/cg-deploy-kubernetes/blob/staging/manifests/kube2iam-template.yaml

### CODE REVIEW

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: I don't think there's a configuration in which `hostNetwork` would be false. Safe to set it explicitly.

**Comment by mgoodness** [comment]
@icereval You beat me to this one! I'm going to review and suggest a few things I was working with.

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: The `host-ip` switch is only required if kube2iam is managing the iptables rule. I'd wrap it in an if statement based on an `iptables` boolean. Probably want the address to default to `$(HOST_IP)` but allow it to be overridden with a variable.

```
{{- if .Values.host.iptables }}
--host-ip={{ default "$(HOST_IP)" .Values.host.ip }}
{{- end }}
```

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: Definitely want this overridable, for those operators who handle the iptables rule through some other mechanism: `--iptables={{ .Values.host.iptables }}` and set it to `false` since that's the kube2iam default.

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: The privileged `securityContext` is only necessary if managing the iptables rule. Put it in an if statement based on the `iptables` boolean:

```
{{- if .Values.host.iptables }}
  securityContext:
    privileged: true
{{- end }}
```

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: How about `--host-interface={{ default "docker0" .Values.host.interface | quote }}` instead of the if block? Makes it clear that the default interface is `docker0`.

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: What uses this? I've never had to set it.

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: I'd include `--app-port={{ default 8181 .Values.containerPort }}` here, and also set `containerPort: {{ default 8181 .Values.containerPort }}` in the port spec.

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: For maximum flexibility, I'm thinking the role-related switches should be omitted in favor of:

```
{{- range $key, $value := .Values.extraArgs }}
   - --{{ $key }}={{ $value }}
{{- end }}
```

Then, in `values.yaml`:

```
extraArgs:
  base-role-arn: arn:aws:iam::0123456789:role/
  default-role: kube2iam-default
  api-server: ...
  api-token: ...
  ...
```

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: See comment on line 25.

Omit `hostPort` and `name`? Only one port is exposed, and there's no need to forward it...

**Review by mgoodness — CHANGES_REQUESTED** [review]
Good-looking chart; lines up closely with what I was doing. Mostly differences in how arguments are handled.

**Comment by icereval** [comment]
@mgoodness , thanks for the review, hoping to get some free time soon to get things updated.

File: incubator/kube2iam/templates/daemonset.yaml
Comment (icereval) [reply incubator/kube2iam/templates/daemonset.yaml]: Tested w/ another region set and does not seem to have any impact.

**Comment by icereval** [comment]
Thanks for the great input, updates pushed h/t @mgoodness

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: You can safely remove this `if` around `--host-interface`. kube2iam defaults to `docker0` if this argument isn't given, so there's no risk of changing behavior if `.Values.host.interface` isn't defined.

Then (IMHO) move it to the top of the args list for clarity.

File: incubator/kube2iam/templates/daemonset.yaml
Comment (mgoodness) [inline incubator/kube2iam/templates/daemonset.yaml]: Same as above. No `if` needed because kube2iam listens on `8181` by default. No risk in passing the argument regardless of whether `.Values.containerPort` is defined.

This one can move up in the args list, too. Again, IMHO.

**Review by mgoodness — COMMENTED** [review]
I would use this chart with no further changes. Just two last (arguably stylistic) suggestions.

Will defer to the Helm team for feedback re: docs.

**Comment by mgoodness** [comment]
LGTM; great chart.

**Comment by prydonius** [comment]
@k8s-bot ok to test

**Comment by lachie83** [comment]
@k8s-bot ok to test

**Comment by viglesiasce** [comment]
I dont have an AWS account to test in can someone else bear the torch of manual/UX testing this?  @mgoodness  @gtaylor

**Comment by mgoodness** [comment]
I'll take it. Should be able to spend some time on it this weekend.

**Comment by fiunchinho** [comment]
May I help somehow to get this merged?

**Comment by viglesiasce** [comment]
Hey @fiunchinho! Have you tested it out? Does it work the way you expect?

**Comment by mgoodness** [comment]
Argh, sorry, I've been bogged down with EOY-type stuff. I'll try to give it some attention this week.

**Comment by mgoodness** [comment]
This LGTM and should be good for `stable`. There is a newer version (0.3.0) of kube2iam, but there are some issues that need be addressed before packaging it.

Sorry again about the delay!

**Review by viglesiasce — CHANGES_REQUESTED** [review]
@icereval can you move the chart to the stable folder. Once that is complete we can merge this in

**Comment by icereval** [comment]
sure, i'll try to get to this over the weekend

**Review by viglesiasce — APPROVED** [review]

**Pull request merged**
